baybe/recommenders/pure/bayesian/botorch.py

```diff
@@ -151,31 +170,174 @@ def _recommend_continuous(
                 f"acquisition functions for batch sizes > 1."
             )
 
+        if len(subspace_continuous.constraints_cardinality):
+            points, _ = self._recommend_continuous_with_cardinality_constraints(
+                subspace_continuous,
+                batch_size,
+            )
+        else:
+            points, _ = self._recommend_continuous_without_cardinality_constraints(
+                subspace_continuous,
+                batch_size,
+            )
+
+        # Return optimized points as dataframe
+        rec = pd.DataFrame(points, columns=subspace_continuous.param_names)
+        return rec
+
+    def _recommend_continuous_with_cardinality_constraints(
+        self,
+        subspace_continuous: SubspaceContinuous,
+        batch_size: int,
+    ) -> tuple[Tensor, Tensor]:
+        """Recommend from a continuous search space with cardinality constraints.
+
+        Args:
+            subspace_continuous: The continuous subspace from which to generate
+                recommendations.
+            batch_size: The size of the recommendation batch.
+
+        Returns:
+            The recommendations.
+            The acquisition values.
+
+        Raises:
+            RuntimeError: If the continuous search space has no cardinality constraint.
+        """
+        import torch
+
+        if not subspace_continuous.constraints_cardinality:
+            raise RuntimeError(
+                f"This method expects a subspace object with constraints of type "
+                f"{ContinuousCardinalityConstraint.__name__}. For a subspace object "
+                f"without constraints of type"
+                f" {ContinuousCardinalityConstraint.__name__}, use method"
+                f"{self._recommend_continuous_without_cardinality_constraints.__name__}."  # noqa
+            )
+
+        acqf_values_all: list[Tensor] = []
+        points_all: list[Tensor] = []
+
+        def append_recommendation_for_inactive_parameters_setting(
+            inactive_parameters: Collection[str],
+        ):
```

Comment:
```suggestion
        ) -> None:
```

Reply:
stuff has been completely refactored / optimized –> thread is obsolete. Will prepare the new version to review shortly and ping you